File: src/main/java/org/c3lang/intellij/psi/C3ModuleNamePsiElement.java
package org.c3lang.intellij.psi;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import java.util.Objects;

public interface C3ModuleNamePsiElement extends C3PsiElement
{
	@Nullable ModuleName getModuleName();

	default boolean isSameModule(@NotNull C3FullyQualifiedNamePsiElement other)
	{
		return Objects.equals(getModuleName(), other.getModuleName());
	}

	default @NotNull String textToInsert(@Nullable ModuleName imported, @NotNull C3FullyQualifiedNamePsiElement element)
	{
		if (isSameModule(element) || element.getModuleName() == null)
			return element.getFqName().getName();

		if (imported != null)
		{
			String relativePath = imported.relativePathTo(element.getModuleName());
			if (relativePath != null)
			{
				if (relativePath.isEmpty()) return element.getFqName().getSuffixName();
				return relativePath + "::" + element.getFqName().getName();
			}
		}

		return element.getFqName().getFullName();
	}
}

File: src/main/java/org/c3lang/intellij/psi/C3ModuleNamePsiElement.kt
package org.c3lang.intellij.psi

import com.intellij.psi.util.parentOfType

interface C3ModuleNamePsiElement : C3PsiElement {
    val moduleName: ModuleName?

    fun isSameModule(other: C3FullyQualifiedNamePsiElement): Boolean {
        return this.containingFile.name == other.containingFile.name && moduleName == other.moduleName
    }

    fun isImported(other: C3FullyQualifiedNamePsiElement): Boolean {
        val moduleDefinition = checkNotNull(parentOfType<C3ModuleDefinition>(true))

        return other.moduleDefinition == moduleDefinition || moduleDefinition.imports.contains(other.moduleName)
    }

    fun textToInsert(imported: ModuleName?, element: C3FullyQualifiedNamePsiElement): String = when {
        isSameModule(element) || imported == null -> element.fqName.name
        isImported(element) || imported == element.moduleName -> element.fqName.suffixName
        else -> element.fqName.fullName
    }
}